Track batch

cruzanmo opened the pull request:
This PR adds support for batch tracking with the method `metrics.track_batch`. This method takes `options` and optional `callback` arguments. The benefit of `track_batch` over `import_batch` is that events reported with `track_batch` will appear in mixpanel's "live view". The limitation of `track_batch` is that the events must be less than 5 days old.

Both `track_batch` and `import_batch` call the `send_batch_requests` function.

One breaking change is that the import_batch method calls the callback with `null` rather than an empty array if there are no errors. This change is useful as it follows standard node callback format, which makes it friendly to methods like `bluebird.promisify`.

Also fixes #114 

`metrics.send_request` has been refactored to handle both GET and POST requests. In this refactor the arguments have been reduced to `options` and `callback`. This is a breaking change to the `send_request` method, which is exposed, but is not documented in the readme. I am assuming that `send_request` is exported only so that it could be easily tested, so users should not need to update their code.

cruzanmo commented:
@tdumitrescu I realize this is a large PR, but any chance of it getting reviewed?

tdumitrescu commented:
Yeah thanks for this! Will try to get to it this week.

tdumitrescu commented on lib/mixpanel-node.js:
```diff
@@ -151,6 +240,128 @@ var create_client = function(token, config) {
+     * @param {Function} handler
+     * @param {Function} callback
+     */
+    var asyncAll = function(requests, handler, callback) {
```
Nice. Any day now we'll get around to dropping support officially for any node.js <v4.2 and can just use Promises natively. In the meanwhile, would you mind adding some basic tests for this helper? Doesn't have to be comprehensive, just enough to give some confidence in its behavior.

tdumitrescu commented on lib/mixpanel-node.js:
```diff
@@ -151,6 +240,128 @@ var create_client = function(token, config) {
+     * @param {object}      options
+     * @param {[{}]}        options.event_list              array of event objects
+     * @param {string}      options.endpoint                e.g. `/track` or `/import`
+     * @param {number}      [options.max_async_requests]    limits concurrent async requests over the network, defaults to all async
```
Good idea. Let's just call this `max_concurrent_requests` though. More importantly, none of the tests for `import_batch` and `track_batch` exercise this new option explicitly right now.

guillegette commented:
Came here to see if anyone was already implementing this. Glad to see it happening.

guillegette commented:
Sorry guys, any chance we can get this merge ? Thanks

cruzanmo commented:
@guillegette I will take a look this week to make changes per @tdumitrescu's comments.

cruzanmo commented:
@tdumitrescu per your comments, I refactored send_request and send_post_request to make it more dry, renamed max_concurrent_requests, and added basic tests for the async helper.

Arekusei commented:
@cruzanmo @tdumitrescu  Will this pull request merged or not?

tdumitrescu commented:
Yes @Arekusei, ETA sometime this week. Thanks for your patience.

Arekusei commented:
@cruzanmo As I see this pull request add support batch requests for track endpoint, but not for engage endpoint? 
https://mixpanel.com/help/reference/http#batch-requests

tdumitrescu commented:
This will be released in v0.7.0 today with a few changes, the most important of which are:
- the `track_batch()` api has been normalized to match that of `import_batch()`, so the first argument is a required event list, rather than an options object
- the `async_all` util has been extracted into a `utils` module rather than exposed on the mixpanel lib object

Thanks for the help!
